### PR TITLE
Renew Blockstack IDs

### PR DESCRIPTION
**Pull request by yknl**
This PR adds name expiry information and basic name renewal feature. Full on-chain names that have less than 60 days until expiration will see a renew button beside it. The renewal transaction will be funded by the built-in BTC wallet.

This is based on a [previous PR](https://github.com/blockstack/blockstack-browser/pull/1924) from @friedger to display expiry information in the all profiles screen.

Renew UI in all profiles screen
image

image

Subdomains do not need to be renewed
image

## Testing notes
1. Find a name that has less than 60 days until expiration.
2. Go to the all profiles screen.
3. Click renew.
4. Enter password and press renew.
5. Wait for transaction to confirm x 10.
6. Verify that domain name now has many more days until expiration.

### CODE REVIEW

**Comment by hstove** [comment]
> Find a name that has less than 60 days until expiration.

This has to be my name? I don't have any names that are that old.

**Comment by yknl** [comment]
You can also test it by making a temporary change, making `60` days limit `600`.

https://github.com/blockstack/blockstack-browser/blob/4f14ba7800b676aacad7c0839ff0b79adfffc922/app/js/profiles/components/IdentityItem.js#L154

**Comment by hstove** [comment]
I would also recommend some 'confirm' step before submitting the actual TX. I had no idea what the cost was going to be

**Comment by markmhx** [comment]
@yknl were you able to incorporate @hstove's suggested changes? They sound helpful to me. Otherwise, this looks functionality like a good addition to me. And I'm wondering if @timstackblock should QA before we go live.

**Comment by yknl** [comment]
@hstove is going to take a stab at improving the UX when he gets a chance

**Comment by timstackblock** [comment]
@markmhx I can take a look is this issue pulled into a zenhub board and ready for QA, I dont see it anywhere?

**Comment by markmhx** [comment]
@timstackblock It sounds from @yknl that @hstove needs to make some UX changes then he'll move this from "In Progress" to "Review / QA" on ZenHub.

**Comment by timstackblock** [comment]
Sure no problem just @ me when you are ready and I will jump right on it.

**Comment by diwakergupta** [comment]
@hstove @markmhx any update here? More and more people are going to run into name expiration.

**Comment by markmhx** [comment]
@diwakergupta I've made a note to discuss this during our weekly kick-off tomorrow.

However, iirc we last decided (over the summer) to keep this effort relatively de-prioritized despite the expectation that names will continue to expire since we don't have a reliable way to reach out to users prompting them to renew (e.g. via email). As such, any UX improvements to make renewals easier will have minimal impact.

**Comment by moxiegirl** [comment]
I've assigned this to myself so I know to write up some documentation for this.

**Comment by moxiegirl** [comment]
s0rry

**Comment by markmhx** [comment]
@hstove Can you provide an estimate of how much time each of the UX improvements you've suggested above would take to implement as part of this PR before we release?

@friedger (or was it @yknl?) appears to have revealed the "Update zone file" and "Transfer name" pages with his addition of a gear icon next to each Blockstack ID on the "Identity" page. I hadn't even seen these before and am a bit concerned that we're making this release more complex by giving those visibility (in addition to adding inline expiration info and renewal option).

Those pages need at least a bit of visual clean-up given that padding between elements is lacking and the "Back" button is strangely placed:

Screen Shot 2019-11-08 at 13 56 40
Screen Shot 2019-11-08 at 13 56 43

How do we feel about linking to these pages and supporting them as part of this release vs. removing the gear option that leads to them? @moxiegirl would these pages increase the surface area for documentation we need to support for this release?

Finally, @yknl is there any way that @timstackblock and others of us could test the "Renew" option here on staging with a test ID set to expire soon – and without making code changes? Mine, for example, expires in 170 days. 

Would it hurt to allow anyone to renew at any time i.e. remove the time limit?

**Comment by moxiegirl** [comment]
@markmhx I can add a new page to the documentation.

Question: is *update* the same a *renew*? That is not clear. If we intend to do this, the *Identity* page would need a sightly, wouldn't it?  I'm not sure I understand how a user navigates to these options.

**Comment by markmhx** [comment]
@moxiegirl By "update" do you mean within the context of "Update zone file" (per [my screenshots above](https://github.com/blockstack/blockstack-browser/pull/1938#issuecomment-551752902)) or something else exactly?

If so, then updating one's zonefile isn't the same as renew. I believe the user can navigate to "Update zone file" and "Transfer name" only via the new gear icon next to a name on the "More" page accessed from "Identity". But I'm not 100% sure.

**Comment by friedger** [comment]
Or guessing the url

This UI only affects the owner of names, not subdomains

Mark Hendrickson <notifications@github.com> schrieb am Mo., 11. Nov. 2019,
12:23:

**Comment by moxiegirl** [comment]
@markmhx Given the title of this ticket and the description contains the term "renewal feature", I expected to find the word *renew* in the interface. I don't see it.

The UI is problematic, users aren't aware of the "zone file" concept.  That might be intentional;  to write any docs, it would be great to have a conversation to better understand the intention here.

**Comment by markmhx** [comment]
The "Renew" option is here to the right of the new gear icon:

64041407-62056b80-cb2d-11e9-9c0b-ed5d888ccf18

I'm also questioning whether we should introduce the zone file concept here and thinking we should probably remove for this release, unless @friedger or @yknl you feel strongly about including?

**Comment by moxiegirl** [comment]
@markmhx so I take it the new dialog --- or is it a page.  Any chance we could get a list of 100 or so IDs to test with? It is a bit of a pain to have to pull the code and rebuild it to test.

I think we should look at making that *Renew* option more prominent.  

I'll have to test this; if we can tell when an ID needs to be renewed to *show* the button why not simply have a date or a countdown with each id?

**Comment by markmhx** [comment]
You should be able to find the staging site for review upon selecting "Details" next to `deploy/netflix` on this PR below (after expanding "Show all checks"). 👇🏻

Per [my comment above](https://github.com/blockstack/blockstack-browser/pull/1938#issuecomment-551752902), we'll have to see if the others here can provide a way for us to test this more easily. And I agree we might just want to show the date or countdown next to all IDs, not just the ones expiring soon.

**Comment by yknl** [comment]
@markmhx Agree that the zone file page will require UI touch up. I'm fine with leaving that out of the next release if we don't have time to improve the UI and provide relevant documentation.

>Finally, @yknl is there any way that @timstackblock and others of us could test the "Renew" option here on staging with a test ID set to expire soon – and without making code changes? Mine, for example, expires in 170 days.

>@markmhx so I take it the new dialog --- or is it a page. Any chance we could get a list of 100 or so IDs to test with? It is a bit of a pain to have to pull the code and rebuild it to test.

You can go directly to the renew page by entering the path `/profiles/0/renew` in the URL bar. The `0` is the index indicating the 1st ID in your list of IDs. You can technically renew an ID at any time, it'll just reset the expiration.

**Comment by timstackblock** [comment]
Sure I can test it I am OOO today but can look at it tomorrow

**Comment by moxiegirl** [comment]
@yknl Thanks. I knew how to get to the test page, it was having an id to test with that appeared to require a rebuild.

**Comment by moxiegirl** [comment]
- "first class identity" is what we want to call this thing created sub to someone's Blockstack id.
- If you want to verify that the work make sure the name goes through.
- Renew transaction new block will move out.

**Comment by yknl** [comment]
> first class identity is what we want to call this thing created sub to someone's Blockstack id.

I'm not sure we've settled on this, it seems like a larger product decision. I recommend not emphasizing this for now.

> For non .id renewals users must use the command line

For renewing any ID not in the `.id` namespace, you will need to use the Blockstack CLI tool

> If you do not renew it someone else can renew the name in the grace priod

No, only the owner can renew the name during the grace period.

> If you want to verify that the work make sure the name goes through.
Renew transaction new block will move out.

For internal testing, you can check that the `expire_block` of the name has changed to a later block after a renewal transaction has been confirmed.
https://core.blockstack.org/v1/names/blockstacker.id

**Comment by hstove** [comment]
Changes I've made:

First, there is a pre-step to get the estimated price, before you actual send the TX:

image

image

Then, I added a note that it might take 2 hours to process, and also disabled the "renew" button after it went through.

image

**Comment by hstove** [comment]
For reviewers:

First, you need a "first class domain", as we called it, aka a name like `myname.id`. Visit the 'identities' page, then click 'more'. Then, you should see a message around how long until it expires, with a renew button next to it, as shown earlier in this ticket. Click the renew button to go to this page.

You need BTC to register a name and renew it. Reach out to me if you need some for testing purposes.

For testing, I've changed the logic to allow you to renew your name even if it's 2 years away. Previously, we set that to 60 days. We should consider what we want that threshold to actually be when we release it.

File: app/js/profiles/AllProfilesPage.js [hunk 1]
```diff
@@ -224,8 +228,8 @@ class AllProfilesPage extends Component {
             </div>
             <div className="row m-t-20">
               <p className="col form-text text-muted">
-                Have you recovered and are missing IDs? Just add them
-                back by using the "Add another ID" for each ID.
+                Have you recovered and are missing IDs? Just add them back by
```
Comment (moxiegirl) [inline app/js/profiles/AllProfilesPage.js]: What does _recovered_ mean here? Recovered from what? Or is it possible to both recover and be missing an ID?  Maybe you want to say:

> Are you missing IDs after restoring the browser? Use  *Add another ID* for each ID you want to add back.

File: app/js/profiles/RenewNamePage.js [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+      return 'Renewal Completed'
+    }
+    if (renewInProgress) {
+      return 'Generating renewal transaction...'
+    }
+    if (estimate) {
+      return 'Renew'
+    }
+    if (estimateInProgress) {
+      return 'Generating estimated cost'
```
Comment (moxiegirl) [inline app/js/profiles/RenewNamePage.js]: _Estimating the purchase price. Price is subject to fluctuations in the price of BTC._

File: app/js/profiles/RenewNamePage.js [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+                name="password"
+                label=""
+                placeholder="Password"
+                type="password"
+                required
+              />
+              {this.state.error ? <div className="alert alert-danger">{this.state.error}</div> : null}
+              {success && <div className="alert alert-success">
+                <p>{success}</p>
+                <p style={{ marginBottom: 0 }}>Your renewal might take up to two hours to process</p>
```
Comment (moxiegirl) [inline app/js/profiles/RenewNamePage.js]: > Processing your renewal can take up to two hours.

If you don't like the rephrase, at least take a period for consistency

File: app/js/profiles/TransferNamePage.js [hunk 4]
```diff
@@ -160,18 +177,22 @@ class TransferNamePage extends Component<Props, State> {
...
+                  <Alert
+                    key={index}
+                    message={alert.message}
+                    status={alert.status}
+                  />
+                ))}
                 <p>
-                Transfer ownership is an advanced feature. It requires broadcasting a
-                transaction on Bitcoin network and costs Bitcoin.
+                  Transfer ownership is an advanced feature. It requires
```
Comment (moxiegirl) [inline app/js/profiles/TransferNamePage.js]: > Transferring the ownership of a name is an advanced feature. Ownership transfer requires broadcasting a transaction on the Bitcoin network; Transactions cost Bitcoin (BTC).

or

> Transferring the ownership of a name is an advanced feature. Ownership transfer requires modifying a name's zone file and broadcasting an update transaction with this change on the Bitcoin network. You must pay for the transaction costs in Bitcoin (BTC).

File: app/js/profiles/ZoneFilePage.js [hunk 5]
```diff
@@ -114,7 +108,10 @@ class ZoneFilePage extends Component {
           disabled: false
         })
       } else if (updateState.broadcasting) {
-        this.updateAlert('success', 'Broadcasting zone file update transaction...')
+        this.updateAlert(
+          'success',
+          'Broadcasting zone file update transaction...'
```
Comment (moxiegirl) [inline app/js/profiles/ZoneFilePage.js]: > Broadcasting name transfer update transaction.

We didn't previously mention the zone file. Are users actually likely to understand what a zone file is? If you use the text I gave you about the zone file up above, then STET this.

File: app/js/profiles/ZoneFilePage.js [hunk 6]
```diff
@@ -176,21 +186,22 @@ class ZoneFilePage extends Component {
...
+                    key={index}
+                    message={alert.message}
+                    status={alert.status}
+                  />
+                ))}
                 <p>
-                Updating your zone file is an advanced feature that can break
-                your Blockstack name and profile. It requires broadcasting a
-                transaction on Bitcoin network and costs Bitcoin.
+                  Updating your zone file is an advanced feature that can break
```
Comment (moxiegirl) [inline app/js/profiles/ZoneFilePage.js]: Why didn't we mention "breaking" above? Maybe just go for consistency and stick with this text:

> Transferring the ownership of a name is an advanced feature. Ownership transfer requires modifying a name's zone file and broadcasting an update transaction with this change on the Bitcoin network. You must pay for the transaction costs in Bitcoin (BTC).

or if we should mention this, this is clearer.

> Transferring the ownership of a name is an advanced feature. Ownership transfer requires modifying a name's zone file and broadcasting an update transaction with this change on the Bitcoin network.  Further, transfers can fail and leave your name in a state where you can no longer use it. You must pay for the transaction costs in Bitcoin (BTC) regardless of whether the transaction succeeds or fails.

File: app/js/profiles/ZoneFilePage.js [hunk 7]
```diff
@@ -222,8 +233,8 @@ class ZoneFilePage extends Component {
                         onChange={this.onToggle}
                         type="checkbox"
                       />
-                      I understand this could break my Blockstack
-                      name and will cost me money.
+                      I understand this could break my Blockstack name and will
+                      cost me money.
```
Comment (moxiegirl) [inline app/js/profiles/ZoneFilePage.js]: > Yes, I understand I am requesting an ownership transfer of name NAMEHERE. Further, I understand that the transfer can fail leaving my name unavailable. Regardless of whether the owner transfer transaction fails or succeeds, I am required to pay transaction fees in Bitcoin (BTC).

File: app/js/profiles/store/registration/actions.js [hunk 8]
```diff
@@ -224,7 +289,7 @@ const registerName = (
             'apps may be unusable until you do.'
           if (res.status === 409) {
             message =
-              "Sorry, it looks like we weren't able to process your name registration. Please contact us at support@blockstack.org for help. Some apps may be unusable until you register an ID."
+              'Sorry, it looks like we weren\'t able to process your name registration. Please contact us at support@blockstack.org for help. Some apps may be unusable until you register an ID.'
```
Comment (moxiegirl) [inline app/js/profiles/store/registration/actions.js]: >   'Sorry, it looks like the blockchain failed to process your name registration. Please contact us at support@blockstack.org for help. Some apps may be unusable until you register an ID.'

This ^^^ is more decentralized language "we were able to process" makes it look like a central authority.

File: app/js/profiles/store/registration/actions.js [hunk 9]
```diff
@@ -251,7 +316,7 @@ const registerName = (
           'apps may be unusable until you do.'
         if (e.status === 409) {
           message =
-            "Sorry, it looks like we weren't able to process your name registration. Please contact us at support@blockstack.org for help. Some apps may be unusable until you register an ID."
+            'Sorry, it looks like we weren\'t able to process your name registration. Please contact us at support@blockstack.org for help. Some apps may be unusable until you register an ID.'
```
Comment (moxiegirl) [inline app/js/profiles/store/registration/actions.js]: > 'Sorry, it looks like the blockchain failed to process your name registration. Please contact us at support@blockstack.org for help. Some apps may be unusable until you register an ID.'

File: app/js/profiles/RenewNamePage.js (hunk 2 above)
Comment (hstove) [reply app/js/profiles/RenewNamePage.js]: This is the button copy - I think that's too long to put in a button. When the estimate is generated, it says "Renewing your name will cost roughly XXX BTC"

**Comment by hstove** [comment]
Thanks @moxiegirl , great feedback! I've updated the copy to what you've suggested.

**Pull request merged**